### PR TITLE
Windows - updates for mswin builds

### PR DESCRIPTION
**Pull request by MSP-Greg**
Updates for mswin builds, no 'bugs' involved.  Commit info:

1. '[package.json and yarn.lock](https://github.com/ruby/setup-ruby/commit/55996aeeb97e7655ea5afacd23e44b6600db10d3)' - 'housekeeping'
2. '[Windows mswin - update addVCVARSEnv](https://github.com/ruby/setup-ruby/commit/c3a022e6a96d4f8c0ae1cf414623a5f4c19e57ec)' - Currently the 'vcvars.bat' file's path was found via a switch/select/case statement that hard-coded the Actions windows image string.  A PowerShell module ([vssetup.powershell](https://github.com/microsoft/vssetup.powershell)) is available for finding information about Visual Studio installs, and it is installed on all Actions Windows runners.  Use the module so hard-coded image strings aren't needed.
3. '[Windows mswin - use ucrt64 tools instead of mingw64](https://github.com/ruby/setup-ruby/commit/58e1d35f716c6bd65bb87f52bc3b4d7b956e8b22)' - Many repos need bash commands, and also use build utilities like bison or ragel.  Currently, mswin builds are using the mingw64 tools, switch to the ucrt64 tools.
4. '[Actions - add windows-2022 mswin job](https://github.com/ruby/setup-ruby/commit/31be906aaecd36952ac494328a126f7e2c664495)'

### CODE REVIEW

**Comment by MSP-Greg** [comment]
Original PR used a PowerShell module known as [vssetup.powershell](https://github.com/microsoft/vssetup.powershell), which contains quite a bit of functionality.  But, running it took anywhere from 14 to 20 sec in one run.  Too long.

Windows images also contain a utility named [vsWhere](https://github.com/Microsoft/vswhere), which is a compiled exe.  It takes around a second to run.  Last commit uses it...

File: windows.js [hunk 1]
```diff
@@ -199,24 +199,17 @@ async function setupMSWin() {
...
-      throw new Error(`Unknown Windows Image: ${virtualEnv}`)
+  let cmd = 'vswhere -latest -property installationPath'
+  let vcVars = `${cp.execSync(cmd).toString().trim()}\\VC\\Auxiliary\\Build\\vcvars64.bat`
+
+  if (!fs.existsSync(vcVars)) {
+    throw new Error(`Missing vcVars file: ${vcVars}`)
   }
   core.exportVariable('VCVARS', vcVars)
 
+  cmd = `cmd.exe /c ""${vcVars}" && set"`
```
Comment (eregon) [inline windows.js:210]: Are the double `"` intended here?

**Review by eregon — APPROVED** [review]
Looks great, thank you and sorry for the review delay

File: windows.js (hunk 1 above)
Comment (MSP-Greg) [reply windows.js:210]: Yes.  There are two pairs, an outer and an inner.  Kind of a pita when Actions uses PowerShell Core for the default script processor, but almost all languages (like Ruby & node.js ) shell out using cmd script.  The two have different styles for handling embedded quotes in commands...

**Comment by MSP-Greg** [comment]
> sorry for the review delay

No problem.  OSS is just a part of our lives...

Do you have any opinion about whether we/I should build the mswin Ruby with OpenSSL 3.0?  Microsoft's vcpkg 'system' has updated to 3.0 from 1.1.  I've also got all the packages needed to build mswin in the [MSP-Greg/setup-msys2-gcc](https://github.com/MSP-Greg/setup-msys2-gcc/releases/tag/msys2-gcc-pkgs) release.  Hence, it could be installed by setup-ruby just like the mingw and ucrt archive files.

I think I'm aware of most of the repos it would affect, and most already work with OpenSSL 3.0...

**Comment by eregon** [comment]
> Do you have any opinion about whether we/I should build the mswin Ruby with OpenSSL 3.0?

I think all GitHub-hosted runners currently use openssl 1.1, not 3, so probably best to stick to 1.1 while it's the case (unless that's problematic for mswin).

**Comment by MSP-Greg** [comment]
GitHub has announced that Ubuntu 22.04 will be available as a preview image some soon.  I believe it uses OpenSSL 3.

I'm running it locally, and building/using Ruby master, even added gcc 12.

Also, the current release of Microsoft's vcpkg system [2022.04.12](https://github.com/microsoft/vcpkg/blob/2022.04.12/ports/openssl/vcpkg.json) builds OpenSSL 3, so sometime soon the Actions Windows runners would build a mswin package with 3.

**Comment by MSP-Greg** [comment]
One thing I haven't mentioned.  mswin builds use a different set of packages than either mingw or ucrt.  In the past, it was kind of a mishmash of sources.  Now, [microsoft/vcpkg](https://github.com/microsoft/vcpkg) has all the packages needed to build mswin (the bison and ragel utilities from MSYS2 work with mswin).

Just last month vcpkg updated OpenSSL from 1.1.1 to 3.0 ([diff](https://github.com/microsoft/vcpkg/commit/407bc79b391a8e30866c95294babca5b111f6006#diff-cb9263a1f84e21b83a97ea8a1bcef649208f2cedafbeea9a50002cdacfca3ed9R3)).  So, a new installation of vcpkg builds OpenSSL 3.0.2.  Packages from vcpkg can be 'exported', so I've also got an archive file of all the packages needed to build mswin.

**Comment by eregon** [comment]
Sounds good. Yeah then it sound like using vcpkg packages is the best for `mswin`.

**Pull request merged**